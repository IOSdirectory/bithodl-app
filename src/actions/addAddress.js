import uuid from 'uuid';

export const ADD_ADDRESS_REQUEST = 'ADD_ADDRESS_REQUEST';
export const ADD_ADDRESS_SUCCESS = 'ADD_ADDRESS_SUCCESS';

export const addAddressRequest = () => {
  return {
    type: ADD_ADDRESS_REQUEST
  };
};

export const addAddressSuccess = (address) => {
  return {
    type: ADD_ADDRESS_SUCCESS,
    address
  };
};

export const addAddress = (address) => {
  return (dispatch) => {
    dispatch(addAddressRequest());
    address.id = address.id || uuid.v1();
    dispatch(addAddressSuccess(address));

    return Promise.resolve(address);
  };
};
