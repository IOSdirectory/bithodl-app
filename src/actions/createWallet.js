import { AsyncStorage } from 'react-native';
import * as Keychain from 'react-native-keychain';
import Mnemonic from 'bitcore-mnemonic';
import toUnixTimestamp from '../utils/toUnixTimestamp';

export const CREATE_WALLET_REQUEST = 'CREATE_WALLET_REQUEST';
export const CREATE_WALLET_SUCCESS = 'CREATE_WALLET_SUCCESS';
export const CREATE_WALLET_FAILURE = 'CREATE_WALLET_FAILURE';

const KEYCHAIN_USERNAME = 'privateKey';
const KEYCHAIN_SERVICE_NAME = 'bithodl';
const STORAGE_WALLET_KEY = '@Wallet';

const savePrivateKey = (mnemonic) => {
  const username = KEYCHAIN_USERNAME;
  const password = mnemonic.toString();

  return Keychain.setGenericPassword(username, password, KEYCHAIN_SERVICE_NAME);
};

const saveMetadata = (mnemonic) => {
  const createdAt = new Date();
  const hdPrivateKey = mnemonic.toHDPrivateKey();
  const hdPublicKey = hdPrivateKey.hdPublicKey;
  const publicKey = hdPublicKey.publicKey;

  const metadata = {
    createdAt: toUnixTimestamp(createdAt),
    publicKey: publicKey.toString() // A DER hex encoded string.
  };

  const serialized = JSON.stringify(metadata);

  return AsyncStorage.setItem(STORAGE_WALLET_KEY, serialized).then(() => {
    return metadata;
  });
};

const createWalletRequest = () => {
  return {
    type: CREATE_WALLET_REQUEST
  };
};

const createWalletSuccess = (metadata) => {
  return {
    type: CREATE_WALLET_SUCCESS,
    metadata
  };
};

const createWalletFailure = (error) => {
  return {
    type: CREATE_WALLET_FAILURE,
    error
  };
};

export const createWallet = (phrase) => {
  return (dispatch) => {
    dispatch(createWalletRequest());

    return Promise.resolve()
      .then(() => {
        const mnemonic = new Mnemonic(phrase);

        return savePrivateKey(mnemonic)
          .then(() => {
            return saveMetadata(mnemonic);
          })
          .then((metadata) => {
            dispatch(createWalletSuccess(metadata));
            return { metadata, mnemonic };
          });
      })
      .catch((error) => {
        dispatch(createWalletFailure(error));
        throw error;
      });
  };
};
