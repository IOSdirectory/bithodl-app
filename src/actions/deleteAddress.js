export const DELETE_ADDRESS_REQUEST = 'DELETE_ADDRESS_REQUEST';
export const DELETE_ADDRESS_SUCCESS = 'DELETE_ADDRESS_SUCCESS';
export const DELETE_ADDRESS_FAILURE = 'DELETE_ADDRESS_FAILURE';

const deleteAddressRequest = () => {
  return {
    type: DELETE_ADDRESS_REQUEST
  };
};

const deleteAddressSuccess = (address) => {
  return {
    type: DELETE_ADDRESS_SUCCESS,
    address
  };
};

const deleteAddressFailure = (error) => {
  return {
    type: DELETE_ADDRESS_FAILURE,
    error
  };
};

export const deleteAddress = (address) => {
  return (dispatch) => {
    dispatch(deleteAddressRequest());

    return Promise.resolve().then(() => {
      if (!address || !address.id) {
        const error = new Error('Unknown address.');
        dispatch(deleteAddressFailure(error));
        throw error;
      }

      dispatch(deleteAddressSuccess(address));

      return address;
    });
  };
};
