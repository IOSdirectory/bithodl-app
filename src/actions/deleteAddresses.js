import { AsyncStorage } from 'react-native';

export const DELETE_ADDRESSES_REQUEST = 'DELETE_ADDRESSES_REQUEST';
export const DELETE_ADDRESSES_SUCCESS = 'DELETE_ADDRESSES_SUCCESS';
export const DELETE_ADDRESSES_FAILURE = 'DELETE_ADDRESSES_FAILURE';

const STORAGE_ADDRESSES_KEY = '@Addresses';

const deleteAddressesRequest = () => {
  return {
    type: DELETE_ADDRESSES_REQUEST
  };
};

const deleteAddressesSuccess = () => {
  return {
    type: DELETE_ADDRESSES_SUCCESS
  };
};

const deleteAddressesFailure = (error) => {
  return {
    type: DELETE_ADDRESSES_FAILURE,
    error
  };
};

export const deleteAddresses = () => {
  return (dispatch) => {
    dispatch(deleteAddressesRequest());

    return AsyncStorage.removeItem(STORAGE_ADDRESSES_KEY)
      .then(() => {
        dispatch(deleteAddressesSuccess());
      })
      .catch((error) => {
        dispatch(deleteAddressesFailure(error));
        throw error;
      });
  };
};
