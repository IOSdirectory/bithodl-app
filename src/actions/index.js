export * from './addAddress';
export * from './createAddress';
export * from './deleteAddress';
export * from './syncAddress';
export * from './updateAddress';
export * from './readAddresses';
export * from './writeAddresses';
export * from './deleteAddresses';
export * from './syncAddresses';
export * from './createWallet';
export * from './getWallet';
export * from './deleteWallet';
export * from './syncWallet';
export * from './createTransaction';
export * from './sendTransaction';
export * from './handleError';
export * from './dismissError';
export * from './navigateWithReset';
export * from './loadSettings';
export * from './saveSettings';
export * from './resetSettings';
export * from './getMnemonic';
export * from './estimateFee';
