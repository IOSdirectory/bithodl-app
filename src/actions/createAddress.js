import uuid from 'uuid';
import toUnixTimestamp from '../utils/toUnixTimestamp';
import createLockedAddress from '../bitcoin/createLockedAddress';
import { getWallet } from './getWallet';
import { addAddress } from './addAddress';
import { writeAddresses } from './writeAddresses';

export const CREATE_ADDRESS_REQUEST = 'CREATE_ADDRESS_REQUEST';
export const CREATE_ADDRESS_SUCCESS = 'CREATE_ADDRESS_SUCCESS';
export const CREATE_ADDRESS_FAILURE = 'CREATE_ADDRESS_FAILURE';

export const createAddressRequest = () => {
  return {
    type: CREATE_ADDRESS_REQUEST
  };
};

export const createAddressSuccess = (address) => {
  return {
    type: CREATE_ADDRESS_SUCCESS,
    address
  };
};

export const createAddressFailure = (error) => {
  return {
    type: CREATE_ADDRESS_FAILURE,
    error
  };
};

export const createAddress = (unlockDate) => {
  return (dispatch) => {
    dispatch(createAddressRequest());

    return dispatch(getWallet())
      .then((wallet) => {
        const publicKey = wallet.publicKey;
        const p2shAddress = createLockedAddress(publicKey, unlockDate);

        const address = {
          id: uuid.v1(),
          hash: p2shAddress.hash,
          script: p2shAddress.script,
          amount: 0,
          createdAt: toUnixTimestamp(new Date()),
          unlockAt: toUnixTimestamp(unlockDate)
        };

        return dispatch(addAddress(address));
      })
      .then((address) => {
        return dispatch(writeAddresses()).then(() => {
          dispatch(createAddressSuccess(address));
          return address;
        });
      })
      .catch((error) => {
        dispatch(createAddressFailure(error));
        throw error;
      });
  };
};
