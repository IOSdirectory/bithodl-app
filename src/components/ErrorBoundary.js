import React, { Component } from 'react';
import { View } from 'react-native';
import PropTypes from 'prop-types';

import { handleError } from '../actions';
import ErrorModalContainer from '../containers/ErrorModalContainer';

export default class ErrorBoundary extends Component {
  componentDidCatch(error) {
    const dispatch = this.props.dispatch;
    dispatch(handleError(error));
  }

  render() {
    return (
      <View {...this.props} style={this.props.style}>
        {this.props.children}
        <ErrorModalContainer />
      </View>
    );
  }
}

ErrorBoundary.propTypes = {
  style: PropTypes.any,
  children: PropTypes.node,
  dispatch: PropTypes.func
};
