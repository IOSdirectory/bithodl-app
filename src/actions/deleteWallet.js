import { AsyncStorage } from 'react-native';
import * as Keychain from 'react-native-keychain';
import { deleteAddresses } from './deleteAddresses';

export const DELETE_WALLET_REQUEST = 'DELETE_WALLET_REQUEST';
export const DELETE_WALLET_SUCCESS = 'DELETE_WALLET_SUCCESS';
export const DELETE_WALLET_FAILURE = 'DELETE_WALLET_FAILURE';

const KEYCHAIN_SERVICE_NAME = 'bithodl';
const STORAGE_WALLET_KEY = '@Wallet';

const deleteMetadata = () => {
  return AsyncStorage.removeItem(STORAGE_WALLET_KEY);
};

const deleteWalletRequest = () => {
  return {
    type: DELETE_WALLET_REQUEST
  };
};

const deleteWalletSuccess = () => {
  return {
    type: DELETE_WALLET_SUCCESS
  };
};

const deleteWalletFailure = (error) => {
  return {
    type: DELETE_WALLET_FAILURE,
    error
  };
};

export const deleteWallet = () => {
  return (dispatch) => {
    dispatch(deleteWalletRequest());

    // Deletes all generic password keychain entries for `KEYCHAIN_SERVICE_NAME`.
    return Keychain.resetGenericPassword(KEYCHAIN_SERVICE_NAME)
      .then(() => {
        return deleteMetadata();
      })
      .then(() => {
        return dispatch(deleteAddresses());
      })
      .then(() => {
        dispatch(deleteWalletSuccess());
      })
      .catch((error) => {
        dispatch(deleteWalletFailure(error));
        throw error;
      });
  };
};
