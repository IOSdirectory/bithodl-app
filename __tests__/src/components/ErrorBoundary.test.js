import React from 'react';
import { Text } from 'react-native';
import renderer from 'react-test-renderer';

import { HANDLE_ERROR } from '../../../src/actions';
import ErrorBoundary from '../../../src/components/ErrorBoundary';

const dispatchMock = jest.fn();

jest.mock('../../../src/containers/ErrorModalContainer', () => 'ErrorModalContainer');

describe('ErrorBoundary', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <ErrorBoundary dispatch={dispatchMock}>
        <Text>83fd03d9-f631-4662-9432-934df7799582</Text>
      </ErrorBoundary>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  describe('#componentDidCatch(error)', () => {
    it('dispatches an action of type HANDLE_ERROR with the error', () => {
      const error = new Error('69f02833-e701-477f-8af5-3cc99ecdd710');

      const errorBoundary = new ErrorBoundary({
        dispatch: dispatchMock
      });

      errorBoundary.componentDidCatch(error);

      expect(dispatchMock).toHaveBeenCalledWith({
        type: HANDLE_ERROR,
        error
      });
    });
  });
});
