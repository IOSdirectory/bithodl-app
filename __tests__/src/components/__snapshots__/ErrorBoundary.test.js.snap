// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary renders correctly 1`] = `
<View
  dispatch={[Function]}
  style={undefined}
>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
  >
    83fd03d9-f631-4662-9432-934df7799582
  </Text>
  <ErrorModalContainer />
</View>
`;
